internal/controller/prometheus_manager.go

```diff
@@ -0,0 +1,103 @@
+	deployedPodTotal   *prometheus.GaugeVec
+}
+
+var (
+	instasliceMetrics = &InstasliceMetrics{
+		// deployed pod total
+		deployedPodTotal: prometheus.NewGaugeVec(prometheus.GaugeOpts{
+			Name: "instaslice_pod_processed_slices",
+			Help: "Pods that are processed with their slice/s allocation.",
+		},
+			[]string{"node", "gpu_id", "namespace", "podname", "profile"}), // Labels: node, GPU ID, namespace, podname, profile
+		// compatible profiles with remaining gpu slices
+		compatibleProfiles: prometheus.NewGaugeVec(prometheus.GaugeOpts{
+			Name: "instaslice_compatible_profiles",
+			Help: "Profiles compatible with remaining GPU slices in a node and their counts.",
+		},
+			[]string{"profile", "node"}), // Labels: profile, node
+		// total processed slices
+		processedSlices: prometheus.NewGaugeVec(prometheus.GaugeOpts{
+			Name: "instaslice_total_processed_gpu_slices",
+			Help: "Number of total processed GPU slices since instaslice controller start time.",
+		},
+			[]string{"node", "gpu_id"}), // Labels: node, GPU ID
+	}
+)
+
+// RegisterMetrics registers all Prometheus metrics
+func RegisterMetrics() {
+	// Register custom metrics with the global prometheus registry
+	metrics.Registry.MustRegister(instasliceMetrics.compatibleProfiles, instasliceMetrics.processedSlices, instasliceMetrics.deployedPodTotal)
+}
+
+// UpdateGpuSliceMetrics updates GPU slice allocation metrics
+func (r *InstasliceReconciler) IncrementTotalProcessedGpuSliceMetrics(instasliceObj inferencev1alpha1.Instaslice, nodeName, gpuID string, profile string) error {
+	processedSlices, err := r.calculateProfileFitOnGPU(&instasliceObj, profile, gpuID, false)
+	if err != nil {
+		return fmt.Errorf("failed to calculate processed GPU slices for profile %s: %w", profile, err)
+	}
+	instasliceMetrics.processedSlices.WithLabelValues(
+		nodeName, gpuID).Add(float64(processedSlices))
+	return nil
+}
+
+// UpdateGpuSliceMetrics updates GPU slice allocation metrics
+func (r *InstasliceReconciler) UpdateDeployedPodTotalMetrics(nodeName, gpuID, namespace, podname, profile string, size int32) error {
+	if namespace == "" && podname == "" && profile == "" {
+		return fmt.Errorf("[UpdateDeployedPodTotalMetrics] : missing required parameters, no prometheus update needed")
+	}
+	instasliceMetrics.deployedPodTotal.WithLabelValues(
+		nodeName, gpuID, namespace, podname, profile).Set(float64(size))
+	return nil
+}
+
+// UpdateCompatibleProfilesMetrics updates metrics based on remaining GPU slices and calculates compatible profiles dynamically
+func (r *InstasliceReconciler) UpdateCompatibleProfilesMetrics(instasliceObj inferencev1alpha1.Instaslice, nodeName string) error {
+	sortedGPUs := sortGPUs(&instasliceObj)
+	// Iterate over each profile
+	for profileName, migPlacement := range instasliceObj.Status.NodeResources.MigPlacement {
+		if len(migPlacement.Placements) > 0 {
+			totalFit := int32(0)
+			// Iterate over all available GPUs
+			for _, gpuID := range sortedGPUs {
```

Comment:
calculate a fit for only one GPU instead of all GPUs on the host, ie I don't think we need a for loop here.

Reply:
Please add a TODO